### PR TITLE
Rptun framework update and related driver update

### PR DESCRIPTION
**Pull request by CV-Bowen**
## Summary
Author: Bowen Wang <wangbowen6@xiaomi.com>
Date:   Wed Dec 20 10:46:07 2023 +0800

    include/nuttx/rptun/rptun.h: change offset type to uint32_t
    
    Sync the offset data type with resource table defined in OpenAMP
    and Linux.
    
    Signed-off-by: Bowen Wang <wangbowen6@xiaomi.com>

Author: mazhuang <mazhuang@xiaomi.com>
Date:   Tue Aug 20 16:32:38 2024 +0800

    rptun/rptun_ivshmem:add restart cmd to reboot slave
    
    Master can send restart command to slave to reboot the slave core
    
    Signed-off-by: mazhuang <mazhuang@xiaomi.com>
    Signed-off-by: Yongrong Wang <wangyongrong@xiaomi.com>

commit 2e717920c90a48e1f11701506f9bb1a683b35caf
Author: xuxingliang <xuxingliang@xiaomi.com>
Date:   Fri Aug 23 19:44:11 2024 +0800

    rptun: fix memleak on failure
    
    Leak backtrace:
        1 14 2096 9886 0x4318d768 [0x040320744] <romfs_fileconfigure+184> romfs/fs_romfsutil.c:1039
                                                     [0x04031fd3e] <romfs_open+378> romfs/fs_romfs.c:281
                                                     [0x04027fa9e] <file_open+446> vfs/fs_open.c:244
                                                     [0x0402ab986] <rptun_store_open+26> rptun/rptun.c:955
                                                     [0x04034cc88] <remoteproc_load+120> open-amp/lib/remoteproc/remoteproc.c:452
                                                     [0x0402ac8ac] <rptun_dev_start+176> rptun/rptun.c:748
                                                     [0x0402ad038] <rptun_ioctl+416> rptun/rptun.c:618
    
    Signed-off-by: xuxingliang <xuxingliang@xiaomi.com>

commit 09b4804d8c95551aa59037ecdbb68266675306d4
Author: Bowen Wang <wangbowen6@xiaomi.com>
Date:   Wed Aug 14 15:01:42 2024 +0800

    rptun: move rptun cmd definition before the resource table
    
    Because locate the command at the end the resource table is unfriendly
    when we want to support multi virtio devices instead only one virtio
    rpmsg device.
    
    Signed-off-by: Bowen Wang <wangbowen6@xiaomi.com>

commit 2ec7b89977f891de6301a3f39e2a52db52d1a19f
Author: Yongrong Wang <wangyongrong@xiaomi.com>
Date:   Tue Aug 6 11:32:14 2024 +0800

    rptun/rpmsg_virtio: remove chip cmd and reuse the common ones
    
    Add more common command for rptun and rpmsg_virtio frameworks,
    also modify the rptun and rpmsg_virtio driver to use the common
    commands.
    
    Signed-off-by: Yongrong Wang <wangyongrong@xiaomi.com>

commit 11430f7411360af0aac56475b7ac438ab49709cd
Author: Yongrong Wang <wangyongrong@xiaomi.com>
Date:   Mon Aug 5 20:00:12 2024 +0800

    sim_rptun.c: remove sim_rptun_panic
    
    Because we can use the common part implemented in rptun
    
    Signed-off-by: Yongrong Wang <wangyongrong@xiaomi.com>

commit 9f4f470217371f84d0094e3320fc61ecd1785532
Author: Yongrong Wang <wangyongrong@xiaomi.com>
Date:   Fri Jul 26 16:57:58 2024 +0800

    rptun.c/rpmsg_virtio.c: move panic logic from chip to rptun/rpmsg_virtio
    
    Move the panic logic in common places, later we can move more logic to
    the framework instead of having the drivers implement it repeatedly.
    
    Signed-off-by: Yongrong Wang <wangyongrong@xiaomi.com>

commit f2278f5f4addab04cefb1ff8b8035d8ded9f794f
Author: Bowen Wang <wangbowen6@xiaomi.com>
Date:   Tue Jun 11 15:53:15 2024 +0800

    sim/sim_rptun: add 64-bit support for sim_rptun
    
    add remote addrenv to make the da is start from 0, so the uint32_t
    da in resource table can store the correct address
    
    Signed-off-by: Bowen Wang <wangbowen6@xiaomi.com>

commit 884396021c365dc16fa7fd75fcf8fbcad167595b
Author: wangyongrong <wangyongrong@xiaomi.com>
Date:   Wed May 22 16:41:21 2024 +0800

    rptun_ivshmem.c: Replace work queue with wdog
    
    wdog has better performance than work queue
    
    Signed-off-by: wangyongrong <wangyongrong@xiaomi.com>

commit 9d2e70ec950aae107a090f4a08c5079eca414fc2
Author: Bowen Wang <wangbowen6@xiaomi.com>
Date:   Sat Aug 3 18:27:44 2024 +0800

    drivers/rptun/rptun.c: move headrx out of CONFIG_RPTUN_PM
    
    headrx is very convient to check weather current core miss interrupt
    by comparing the headrx with the rx vring avail.idx for slave side or
    rx vring used.idx for master side.
    
    So move headrx out of the CONFIG_RPTUN_PM range.
    
    Signed-off-by: Bowen Wang <wangbowen6@xiaomi.com>

commit 462f7d7513bff0cb06b326d25d2af5d828f29840
Author: Bowen Wang <wangbowen6@xiaomi.com>
Date:   Fri Feb 23 19:57:25 2024 +0800

    rptun: rptun pm and rptun dump support cacheable memory
    
    Should invalidate the memory when the data is located in shared
    memory and write by remote core.
    
    Signed-off-by: Bowen Wang <wangbowen6@xiaomi.com>

commit 7aa295cd26c4a5f2cc13351d905334cd2bf62db6
Author: Xiang Xiao <xiaoxiang@xiaomi.com>
Date:   Sun Feb 18 20:38:46 2024 +0800

    rptun: Rename rptun_panic_ to rptun_panic
    
    The function name rptun_panic_ is not consistent with other functions
    
    Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>

commit 218530953e8731c54607811c7cf5129d867ec7a7
Author: ligd <liguiding1@xiaomi.com>
Date:   Wed Apr 27 16:58:02 2022 +0800

    rptun: add timeout to wait_tx_buffer callback
    
    Signed-off-by: ligd <liguiding1@xiaomi.com>

commit 62ca344eb2d11e9876b6b87bd217d6a417c576e7
Author: Bowen Wang <wangbowen6@xiaomi.com>
Date:   Wed Sep 18 17:26:16 2024 +0800

    rptun/rptun_dump: remove unused rptun_dump.c
    
    rptun_dump related code has been moved to rptun.c from rptun_dump.c,
    but file rptun_dump.c is not deleted in PR:
    https://github.com/apache/nuttx/pull/11712
    
    So delete this file.
    
    Signed-off-by: Bowen Wang <wangbowen6@xiaomi.com>

commit e8e64e6377abfee5b8df16f76640792fe9fd0755
Author: Bowen Wang <wangbowen6@xiaomi.com>
Date:   Wed Sep 18 17:20:56 2024 +0800

    rptun: BUG fix, should not destory the semaphore twice
    
    This BUG is introduced in PR: https://github.com/apache/nuttx/pull/13172
    
    Signed-off-by: Bowen Wang <wangbowen6@xiaomi.com>

commit f79daeddd4fe90b689169908e9c4d1077b5e2902
Author: ligd <liguiding1@xiaomi.com>
Date:   Mon Jun 20 21:47:20 2022 +0800

    rptun: use detail name for pm wakelock
    
    So we can distinguish the pm wakelock
    
    Signed-off-by: ligd <liguiding1@xiaomi.com>

commit e8ad329319878c6e5ffeeba0b465ef4455c24d6c
Author: Bowen Wang <wangbowen6@xiaomi.com>
Date:   Thu Mar 28 15:16:40 2024 +0800

    drivers/rptun: flush the image memory when read from the file system
    
    Flush the image memory to make sure the remote core access the correct
    image.
    
    Signed-off-by: Bowen Wang <wangbowen6@xiaomi.com>

commit 29c8a585777137cdcb2e025112f62aa145898255
Author: ligd <liguiding1@xiaomi.com>
Date:   Fri Apr 7 23:38:33 2023 +0800

    rptun: add RPTUN_PM_AUTORELAX method.
    
    Signed-off-by: ligd <liguiding1@xiaomi.com>

commit afb0ea8a803f8f45abb24c325b2bc5080aa78d50
Author: ligd <liguiding1@xiaomi.com>
Date:   Thu Jun 16 20:38:51 2022 +0800

    rptun/pm: use pm_wakelock
    
    Signed-off-by: ligd <liguiding1@xiaomi.com>

commit abe9fd4d4bb7034f94950ff837cd6d53578f54ef
Author: ligd <liguiding1@xiaomi.com>
Date:   Thu Jan 5 12:08:04 2023 +0800

    rptun/pm: add check to rptun_pm_callback() incase of start early
    
    Signed-off-by: ligd <liguiding1@xiaomi.com>

commit 95d3611e1c70fc2ace5507884d73b0d5d605d122
Author: ligd <liguiding1@xiaomi.com>
Date:   Sat Apr 30 12:46:27 2022 +0800

    rptun: use local rx virtqueue idx to resolve remote low power
    
    Store the rx virtqueue idx to the local headrx index, and only
    process the data when the rx virtqueue has data to avoid access
    the ram in low power mode.
    
    Signed-off-by: ligd <liguiding1@xiaomi.com>

commit dab9e6664395ae7d332f6abb5e0eb0649f0ef4dd
Author: yanghuatao <yanghuatao@xiaomi.com>
Date:   Thu Aug 10 16:06:57 2023 +0800

    nuttx/dirvers: Add secure rptun file
    
    rptun secure is a rptun driver used for the rpmsg communication
    between REE and TEE.
    
    Signed-off-by: yanghuatao <yanghuatao@xiaomi.com>

## Impact
rptun

## Testing
sim:rpserver, rpproxy and qemu

### CODE REVIEW

**Comment by lupyuen** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13494&type=pullrequests)

## NuttX PR Requirements Review

This PR appears to bundle multiple changes related to the `rptun` driver, including bug fixes, feature additions, and refactoring.  While the individual commit messages provide some details, the overall PR lacks a cohesive summary and impact analysis. 

**Summary:**

* **Why:**  The PR description should clearly state the overall motivation and goals for the collection of changes. 
* **What:** A high-level overview of the functional areas impacted by the changes (e.g., data types, commands, memory management, etc.) is missing.
* **How:**  The PR description should briefly explain how these changes are implemented.

**Impact:**

* The "Impact" section is overly brief. Each bullet point should elaborate on how the changes affect the specified area. For instance:
    * **Is new feature added?**  Yes, the PR introduces a "restart" command and support for 64-bit in `sim_rptun`. A description of these features is needed.
    * **Impact on user?** The PR mentions "Master can send restart command to slave." This needs further explanation on how a user would utilize this command and any potential adaptations required.
    * **Impact on build?** If there are no build process changes, state it explicitly.  If there are, provide details.
    * **Impact on hardware?** The testing section lists "sim:rpserver, rpproxy and qemu." Specify the affected architectures and boards.
    * **Impact on documentation?**  Clarify if documentation updates are needed and if they are included in the PR.
    * **Impact on security?** The PR introduces a `rptun_secure` file. Explain the security implications and any potential risks or mitigations.
    * **Impact on compatibility?** Are there any backward or forward compatibility concerns introduced by the changes?

**Testing:**

* The provided testing logs are placeholders.  Include actual logs demonstrating the functionality before and after the changes.
* The testing section should be more comprehensive, covering each significant change in the PR.

**Overall:**

This PR requires significant improvements to meet the NuttX requirements. Provide a clear and detailed summary, elaborate on the impact of each change, and include actual testing logs to ensure the changes are thoroughly validated.

File: drivers/rptun/Kconfig [hunk 1]
```diff
@@ -12,6 +12,12 @@ menuconfig RPTUN
 
 if RPTUN
 
+config RPTUN_SECURE
+	bool "rptun secure support"
+	default n
+	---help---
+		This is for secure rptun support.
```
Comment (acassis) [inline drivers/rptun/Kconfig:19]: Please include more information here. Why is it considered secure? What is main differences?

File: drivers/rptun/rptun.c [hunk 2]
```diff
@@ -280,6 +370,30 @@ static bool rptun_is_recursive(FAR struct rptun_priv_s *priv)
...
+    {
+      cmd = rptun_cmd->cmd_master;
+      rptun_cmd->cmd_master = 0;
+    }
+
+  switch (RPTUN_GET_CMD(cmd))
+    {
+      case RPTUN_CMD_PANIC:
+        PANIC();
+        break;
```
Comment (pkarashchenko) [inline drivers/rptun/rptun.c:393]: Good to add explicit `default` case or change to `if`

File: drivers/rptun/rptun.c [hunk 3]
```diff
@@ -280,6 +370,30 @@ static bool rptun_is_recursive(FAR struct rptun_priv_s *priv)
   return nxsched_gettid() == priv->tid;
 }
 
+static void rptun_command(struct rptun_priv_s *priv)
```
Comment (pkarashchenko) [inline drivers/rptun/rptun.c:373]:
```suggestion
static void rptun_command(FAR struct rptun_priv_s *priv)
```

File: drivers/rptun/rptun_secure.c [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  int                     irq_trigger;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static const char *rptun_secure_get_cpuname(FAR struct rptun_dev_s *dev);
+static struct
+rptun_rsc_s *rptun_secure_get_resource(FAR struct rptun_dev_s *dev);
```
Comment (pkarashchenko) [inline drivers/rptun/rptun_secure.c:55]:
```suggestion
FAR rptun_rsc_s *rptun_secure_get_resource(FAR struct rptun_dev_s *dev);
```

File: drivers/rptun/rptun_secure.c [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  char                    cpuname[RPMSG_NAME_SIZE + 1];
+  int                     irq_event;
+  int                     irq_trigger;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static const char *rptun_secure_get_cpuname(FAR struct rptun_dev_s *dev);
```
Comment (pkarashchenko) [inline drivers/rptun/rptun_secure.c:53]:
```suggestion
static FAR const char *rptun_secure_get_cpuname(FAR struct rptun_dev_s *dev);
```

File: drivers/rptun/rptun_secure.c [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  .stop              = rptun_secure_stop,
+  .notify            = rptun_secure_notify,
+  .register_callback = rptun_secure_register_callback,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static const char *rptun_secure_get_cpuname(FAR struct rptun_dev_s *dev)
```
Comment (pkarashchenko) [inline drivers/rptun/rptun_secure.c:86]:
```suggestion
static FAR const char *rptun_secure_get_cpuname(FAR struct rptun_dev_s *dev)
```

File: drivers/rptun/rptun_secure.c [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+{
+  FAR struct rptun_secure_dev_s *priv = (FAR struct rptun_secure_dev_s *)dev;
+  return priv->cpuname;
+}
+
+static struct rptun_rsc_s *
```
Comment (pkarashchenko) [inline drivers/rptun/rptun_secure.c:92]:
```suggestion
static FAR struct rptun_rsc_s *
```

File: drivers/rptun/Kconfig (hunk 1 above)
Comment (CV-Bowen) [reply drivers/rptun/Kconfig]: Done

File: drivers/rptun/rptun.c (hunk 2 above)
Comment (CV-Bowen) [reply drivers/rptun/rptun.c:393]: Done

File: drivers/rptun/rptun_secure.c (hunk 4 above)
Comment (CV-Bowen) [reply drivers/rptun/rptun_secure.c]: Done

File: drivers/rptun/rptun_secure.c (hunk 5 above)
Comment (CV-Bowen) [reply drivers/rptun/rptun_secure.c]: Done

File: drivers/rptun/rptun_secure.c (hunk 6 above)
Comment (CV-Bowen) [reply drivers/rptun/rptun_secure.c]: Done

File: drivers/rptun/rptun_secure.c (hunk 7 above)
Comment (CV-Bowen) [reply drivers/rptun/rptun_secure.c]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**